architecture/authentication.adoc

```diff
@@ -0,0 +1,268 @@
+
+==== Session configuration [[config-session]]
+
+This configuration is used by the link:#config-request-handlers[`session`] request handler. 
+
+The following environment variables control the cookie-based session:
+
+* `ORIGIN_OAUTH_SESSION_SECRET`
+** Secret used to encrypt cookie sessions
+** Defaults to a random uuid on each server start
+* `ORIGIN_OAUTH_SESSION_NAME`
+** Name of the cookie used to store the session
+** Defaults to `ssn`
+* `ORIGIN_OAUTH_SESSION_MAX_AGE_SECONDS`
+** Controls the maximum age of a session (sessions auto-expire once a token request is complete)
+** If auto-grant is not enabled, sessions must last as long as the user is expected to take to approve or reject a client authorization request
+** Defaults to 5 minutes
+
+==== Google external identity provider [[config-google]]
+
+This configuration is used by the link:#config-auth-handler[`google`] auth handler.
+
+* `ORIGIN_OAUTH_GOOGLE_CLIENT_ID`
+** The client_id of a client registered with the https://developers.google.com/accounts/docs/OAuth2[Google OAuth provider]
+** It must be authorized to redirect to `MasterPublicAddr`/oauth2callback/google
+* `ORIGIN_OAUTH_GOOGLE_CLIENT_SECRET`
+** The client_secret of a client registered with the Google OAuth provider
+
+==== Github external identity provider [[config-github]]
+
+This configuration is used by the link:#config-auth-handler[`github`] auth handler.
+
+* `ORIGIN_OAUTH_GITHUB_CLIENT_ID`
+** The client_id of a client registered with the https://developer.github.com/v3/oauth/[GitHub OAuth provider]
+** It must be authorized to redirect to `MasterPublicAddr`/oauth2callback/github
+* `ORIGIN_OAUTH_GITHUB_CLIENT_SECRET`
+** The client_secret of a client registered with the GitHub OAuth provider
+
+==== Handle unauthorized client requests [[config-grant-handler]]
+
+To configure how the OAuth server responds to token requests for a client the user has not previously authorized, set the `ORIGIN_OAUTH_GRANT_HANDLER` environment variable. Valid values are:
+
+* `auto`
+** Auto-approve the grant and retry the request
+* `prompt`
+** Prompt the user to approve or deny the grant
+* `deny`
+** Auto-deny the grant and return a failure error to the client
+
+`ORIGIN_OAUTH_GRANT_HANDLER` defaults to `auto`
+
+==== Token generation [[config-token]]
+
+The OAuth server can generate two kinds of tokens. 
+
+Authorize codes are short-lived tokens whose only use is to be exchanged for an access token. Set the `ORIGIN_OAUTH_AUTHORIZE_TOKEN_MAX_AGE_SECONDS` environment variable to control the lifetime of authorize codes. The default lifetime is 5 minutes.
+
+Access tokens are longer-lived tokens that grant access to the API. Set the `ORIGIN_OAUTH_ACCESS_TOKEN_MAX_AGE_SECONDS` environment variable to control the lifetime of access tokens. The default lifetime is 1 hour.
+
+=== Examples
+
+==== Default
+When running `openshift start` with no OAuth environment variables set, these are the default settings:
```

Comment:
Describe what these do.  I know a user could look it up based in this doc, but it's harder to parse.